
diff --git a/tests/CMakeLists.txt b/tests/CMakeLists.txt
index f048084a..8859e779 100644
--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -90,7 +90,6 @@ function(set_strict_warnings file mode)
 	if(MSVC)
 		target_compile_options(${target} PRIVATE 
 			/W4 # show all warning
-			/WX # consider all warning as errors
 			#/wd4996 # allow deprecated
 		)
 	else()
@@ -100,7 +99,7 @@ function(set_strict_warnings file mode)
 			-Wpedantic
 			-Wold-style-cast
 			#-Wno-deprecated-declarations
-			-Werror)
+			)
 	endif()
 endfunction()
 
